linera-rpc/src/grpc/server.rs

```diff
@@ -397,77 +398,113 @@ where
         cross_chain_retry_delay: Duration,
         cross_chain_sender_delay: Duration,
         cross_chain_sender_failure_rate: f32,
-        cross_chain_max_concurrent_tasks: usize,
         this_shard: ShardId,
-        receiver: mpsc::Receiver<(linera_core::data_types::CrossChainRequest, ShardId)>,
+        mut receiver: mpsc::Receiver<(linera_core::data_types::CrossChainRequest, ShardId)>,
     ) {
         let pool = GrpcConnectionPool::default();
-        let max_concurrent_tasks = Some(cross_chain_max_concurrent_tasks);
+        let mut join_set = JoinSet::new();
+        // An entry means that a task is running for the given sender, recipient and request kind.
+        // If it is Some, that means the task is still working on another request, and will pick
+        // up the new one when it is done. Since later requests always supersede earlier ones,
+        // we can always overwrite the entry if a new one comes in.
+        let cross_chain_tasks = Arc::new(Mutex::new(HashMap::new()));
+
+        while let Some((mut cross_chain_request, shard_id)) = receiver.next().await {
+            {
+                let mut guard = cross_chain_tasks.lock().unwrap();
+                match guard.entry(cross_chain_request.sender_recipient_update()) {
+                    Entry::Occupied(mut entry) => {
+                        // A task is already running; it will pick up the new request when done.
+                        entry.insert(Some(cross_chain_request));
+                        continue;
+                    }
+                    Entry::Vacant(entry) => {
+                        entry.insert(None);
+                    }
+                }
+            }
+            // Spawn a task to send the cross-chain query.
+            let shard = network.shard(shard_id);
+            let remote_address = shard.http_address();
+
+            let pool = pool.clone();
+            let nickname = nickname.clone();
+            let cross_chain_tasks = cross_chain_tasks.clone();
+            join_set.spawn(async move {
+                // Send the cross-chain query and retry if needed.
+                if cross_chain_sender_failure_rate > 0.0
+                    && rand::thread_rng().gen::<f32>() < cross_chain_sender_failure_rate
+                {
+                    warn!("Dropped 1 cross-chain message intentionally.");
+                    return;
+                }
```

Comment:
We should be careful to drop the entry here or no new messages will ever be delivered through this channel.

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/3938/commits/9985c0bc8a2b377e75f7d7c44b03d4085627bf3a.